zebra-consensus/src/primitives/halo2/tests.rs

```diff
@@ -0,0 +1,240 @@
+                        zebra_chain::orchard::shielded_data::AuthorizedAction {
+                            action,
+                            spend_auth_sig: <[u8; 64]>::from(a.authorization()).into(),
+                        }
+                    })
+                    .collect::<Vec<_>>()
+                    .try_into()
+                    .unwrap(),
+                binding_sig: <[u8; 64]>::from(bundle.authorization().binding_signature())
+                    .try_into()
+                    .unwrap(),
+            }
+        })
+        .collect();
+
+    for sd in shielded_data {
+        println!(
+            "{}",
+            hex::encode(sd.clone().zcash_serialize_to_vec().unwrap())
+        );
+    }
+}
+
+async fn verify_orchard_halo2_proofs<V>(
+    verifier: &mut V,
+    shielded_data: Vec<ShieldedData>,
+) -> Result<(), V::Error>
+where
+    V: tower::Service<Item, Response = ()>,
+    <V as tower::Service<Item>>::Error: From<tower::BoxError>,
+    <V as tower::Service<Item>>::Error: std::fmt::Debug,
+{
+    let mut async_checks = FuturesUnordered::new();
+
+    for sd in shielded_data {
+        tracing::trace!(?sd);
+
+        let rsp = verifier.ready().await?.call(Item::from(sd));
+
+        async_checks.push(rsp);
+    }
+
+    while let Some(result) = async_checks.next().await {
+        tracing::trace!(?result);
+        result?;
+    }
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn verify_generated_halo2_proofs() {
+    zebra_test::init();
+
+    // These test vectors are generated by `generate_text_vectors()` function.
+    let shielded_data = vectors::ORCHARD_SHIELDED_DATA
+        .clone()
+        .iter()
+        .map(|bytes| {
+            let maybe_shielded_data: Option<zebra_chain::orchard::ShieldedData> = bytes
+                .zcash_deserialize_into()
+                .expect("a valid orchard::ShieldedData instance");
+            maybe_shielded_data.unwrap()
+        })
+        .collect();
+
+    // Use separate verifier so shared batch tasks aren't killed when the test ends (#2390)
+    let mut verifier = Fallback::new(
+        Batch::new(
+            Verifier::new(&VERIFYING_KEY),
+            crate::primitives::MAX_BATCH_SIZE,
+            crate::primitives::MAX_BATCH_LATENCY,
+        ),
+        tower::service_fn(
+            (|item: Item| ready(item.verify_single(&VERIFYING_KEY).map_err(Halo2Error::from)))
+                as fn(_) -> _,
+        ),
+    );
+
+    println!("Display: {}", Halo2Error::Other);
+
+    // This should fail if any of the proofs fail to validate.
+    assert!(verify_orchard_halo2_proofs(&mut verifier, shielded_data)
+        .await
+        .is_ok());
+}
+
+async fn verify_invalid_orchard_halo2_proofs<V>(
+    verifier: &mut V,
+    shielded_data: Vec<ShieldedData>,
+) -> Result<(), V::Error>
+where
+    V: tower::Service<Item, Response = ()>,
+    <V as tower::Service<Item>>::Error: From<tower::BoxError>,
+    <V as tower::Service<Item>>::Error: std::fmt::Debug,
```

Comment:
Minor: same as above:

```suggestion
    <V as tower::Service<Item>>::Error: From<tower::BoxError> + std::fmt::Debug,
```

Reply:
Done in 4681f761